std/socket.zig

```diff
@@ -0,0 +1,676 @@
+            }
+        }
+    }
+
+    pub fn tcp(domain: Domain) SocketError!Socket {
+        return Socket.open(domain, .Stream, .TCP);
+    }
+
+    pub fn udp(domain: Domain) SocketError!Socket {
+        return Socket.open(domain, .DataGram, Proocol.UDP);
+    }
+
+    pub fn unix(socket_type: SocketType) SocketError!Socket {
+        return Socket.open(.Unix, socket_type, .Unix);
+    }
+
+    pub fn bind(self: Socket, addr: *const net.Address) BindError!void {
+        const rc = system.bind(self.fd, &addr.os_addr, @sizeOf(net.OsAddress));
+
+        if (is_posix) {
+            switch (system.getErrno(@intCast(usize, rc))) {
+                0 => return,
+                system.EACCES => return BindError.AccessDenied,
+                system.EADDRINUSE => return BindError.AddressInUse,
+                system.EBADF => unreachable, // always a race condition if this error is returned
+                system.EINVAL => unreachable,
+                system.ENOTSOCK => unreachable,
+                system.EADDRNOTAVAIL => return BindError.AddressNotAvailable,
+                system.EFAULT => unreachable,
+                system.ELOOP => return BindError.SymLinkLoop,
+                system.ENAMETOOLONG => return BindError.NameTooLong,
+                system.ENOENT => return BindError.FileNotFound,
+                system.ENOMEM => return BindError.system.emResources,
+                system.ENOTDIR => return BindError.NotDir,
+                system.EROFS => return BindError.ReadOnlyFilesystem.em,
+                else => |err| return unexpectedError(err),
+            }
+        } else {
+            switch (WSALastError(rc)) {
+                0 => return,
+                system.ERROR.WSAEACCES => return BindError.AccessDenied,
+                system.ERROR.WSAEADDRINUSE => return BindError.AddressInUse,
+                system.ERROR.WSAEINVAL => unreachable,
+                system.ERROR.WSAENOTSOCK => unreachable,
+                system.ERROR.WSAEADDRNOTAVAIL => return BindError.AddressNotAvailable,
+                system.ERROR.WSAEFAULT => unreachable,
+                else => |err| return unexpectedError(@intCast(u32, err)),
+            }
+        }
+    }
+
+    pub fn listen(self: Socket, backlog: u32) ListenError!void {
+        if (is_posix) {
+            const rc = system.listen(self.fd, backlog);
+            switch (system.getErrno(@intCast(usize, rc))) {
+                0 => return,
+                system.EADDRINUSE => return ListenError.AddressInUse,
+                system.EBADF => unreachable,
+                system.ENOTSOCK => return ListenError.FileDescriptorNotASocket,
+                system.EOPNOTSUPP => return ListenError.OperationNotSupported,
+                else => |err| return unexpectedError(err),
+            }
+        } else {
+            const rc = system.listen(self.fd, @intCast(c_int, backlog));
+            switch (WSALastError(rc)) {
```

Comment:
Same thing here, check the return code first.  I think there's a few more places as well.

Reply:
see [here](https://github.com/ziglang/zig/pull/2822#discussion_r305532770).